libnetwork/driver_plugin.py

```diff
@@ -70,13 +96,131 @@ def make_json_error(ex):
 # The API calls below are documented at
 # https://github.com/docker/libnetwork/blob/master/docs/remote.md
 
+# <-- Plugin activation, we activate both libnetwork and ipam plugins -->
 
 @app.route('/Plugin.Activate', methods=['POST'])
 def activate():
-    json_response = {"Implements": ["NetworkDriver"]}
+    json_response = {"Implements": ["NetworkDriver", "IpamDriver"]}
     app.logger.debug("Activate response JSON=%s", json_response)
     return jsonify(json_response)
 
+# <-- IPAM plugin API -->
+
+@app.route('/IpamDriver.GetDefaultAddressSpaces', methods=['POST'])
+def get_default_address_spaces():
+    json_data = request.get_json(force=True)
+    app.logger.debug("GetDefaultAddressSpace JSON=%s", json_data)
+    json_response = {
+        "LocalDefaultAddressSpace": LOCAL_CIDR_ADDR_SP,
+        "GlobalDefaultAddressSpace": GLOBAL_CIDR_ADDR_SP
+    }
+    app.logger.debug("GetDefaultAddressSpace response JSON=%s", json_response)
+    return jsonify(json_response)
+
+
+@app.route('/IpamDriver.RequestPool', methods=['POST'])
+def request_pool():
+    json_data = request.get_json(force=True)
+    app.logger.debug("RequestPool JSON=%s", json_data)
+    address_space = json_data["AddressSpace"]
+    pool = json_data["Pool"]
+    sub_pool = json_data.get("SubPool")
+    options = json_data["Options"]
+    v6 = json_data["V6"]
+
+    # Calico IPAM does not allow you to request Pool or SubPool.
+    assert not pool and not sub_pool, \
```

Comment:
Is assert the right thing to use here? The API defines an error response which we try to confirm to using make_json_error(). Can you confirm that using assert gives the right user experience. This is easy to check manually which you should do in addition to adding a UT.